source/simplex/quadrature_lib.cc

```diff
@@ -263,6 +159,248 @@ namespace Simplex
+    // reasonable equivalent for generic programming.
+    if (dim == 1)
+      {
+        const dealii::QGauss<dim> quad(n_points_1D);
+        this->quadrature_points = quad.get_points();
+        this->weights           = quad.get_weights();
+        return;
+      }
+
+    std::array<double, dim + 1> centroid;
+    std::fill(centroid.begin(), centroid.end(), 1.0 / (dim + 1.0));
+    std::vector<std::vector<std::array<double, dim + 1>>> b_point_permutations;
+    std::vector<double>                                   b_weights;
+
+    // We can simplify the implementation of these quadrature rules
+    // by quite a bit by exploiting symmetry - we do essentially the
+    // same thing for each barycentric coordinate, so we can express
+    // our quadrature rule as permutations of barycentric points
+    // instead of writing things out explicitly.
+
+    // Apply a Barycentric permutation where one point is different.
+    auto process_point_1 = [&](const double a, const double w) {
+      const double                b = 1.0 - dim * a;
+      std::array<double, dim + 1> b_point;
+      std::fill(b_point.begin(), b_point.end(), a);
+      b_point[dim] = b;
+
+      b_weights.push_back(w);
+      b_point_permutations.push_back(all_permutations(b_point));
+    };
+
+    // Apply a Barycentric permutation where two points (in 3D) are different.
+    auto process_point_2 = [&](const double a, const double w) {
+      Assert(dim == 3, ExcInternalError());
+      const double                b = (1.0 - 2.0 * a) / 2.0;
+      std::array<double, dim + 1> b_point;
+      std::fill(b_point.begin(), b_point.end(), a);
+      b_point[dim - 1] = b;
+      b_point[dim]     = b;
+
+      b_weights.push_back(w);
+      b_point_permutations.push_back(all_permutations(b_point));
+    };
+
+    // Apply a Barycentric permutation where three (or four) points
+    // are different (since there are two inputs).
+    auto process_point_3 = [&](const double a, const double b, const double w) {
+      const double                c = 1.0 - (dim - 1.0) * a - b;
+      std::array<double, dim + 1> b_point;
+      std::fill(b_point.begin(), b_point.end(), a);
+      b_point[dim - 1] = b;
+      b_point[dim]     = c;
+
+      b_weights.push_back(w);
+      b_point_permutations.push_back(all_permutations(b_point));
+    };
+
+    if (n_points_1D == 1)
+      {
+        b_point_permutations.push_back({centroid});
+        b_weights.push_back(1.0);
+      }
+    else if (n_points_1D == 2)
+      {
+        // This is WV-4 in 2D and WV-3 in 3D
+        if (dim == 2)
+          {
+            process_point_1(9.1576213509770743e-02, 1.0995174365532187e-01);
+            process_point_1(4.4594849091596489e-01, 2.2338158967801147e-01);
+          }
+        else if (dim == 3)
+          {
+            process_point_1(3.281633025163817e-01, 1.362178425370874e-01);
+            process_point_1(1.080472498984286e-01, 1.137821574629126e-01);
+          }
```

Comment:
could you add
```suggestion
          }
       else
         Assert (false, ExcNotImplemented());
```
here and to the corresponding cases below? This way, if someone for whatever reason tried to do `dim==4`, there would at least be an error message.

Reply:
Would you be OK with me just adding this check at the top of the function? I'd rather write it once there than for every single `n_points_1D` value.

Reply:
I guess that's ok :-)